File: bfl_custom/bfl_custom/page/latest_journal_entry/latest_journal_entry.py
import frappe
from frappe import _

@frappe.whitelist()
def get_latest_journal_entries(account, from_date):

    # ⭐ MAIN QUERY WITH OPENING + RUNNING
    data = frappe.db.sql("""

        WITH opening AS (
            SELECT 
                COALESCE(
                    SUM(jea.debit_in_account_currency 
                      - jea.credit_in_account_currency), 0
                ) AS opening_balance
            FROM `tabJournal Entry Account` jea
            JOIN `tabJournal Entry` je 
                ON je.name = jea.parent
            WHERE
                jea.account = %(account)s
                AND je.posting_date < %(from_date)s
        ),

        main AS (
            SELECT
                je.name AS voucher,
                je.posting_date,
                je.creation,

                cash_line.account AS cash_account,

                GROUP_CONCAT(
                    CONCAT(
                        other_line.account,
                        IF(other_line.party IS NOT NULL,
                            CONCAT(' (', other_line.party, ')'), ''
                        ),
                        ' D:', other_line.debit_in_account_currency,
                        ' C:', other_line.credit_in_account_currency
                    ) SEPARATOR '; '
                ) AS counter_details,

                je.remark,

                CASE je.docstatus
                    WHEN 0 THEN 'Draft'
                    WHEN 1 THEN 'Submitted'
                    WHEN 2 THEN 'Cancelled'
                END AS status,

                je.custom_payment_type,

                cash_line.debit_in_account_currency AS debit,
                cash_line.credit_in_account_currency AS credit,

                SUM(
                    cash_line.debit_in_account_currency
                  - cash_line.credit_in_account_currency
                ) OVER (
                    ORDER BY je.posting_date, je.name
                ) AS running_balance

            FROM `tabJournal Entry` je

            JOIN `tabJournal Entry Account` cash_line
                ON je.name = cash_line.parent

            LEFT JOIN `tabJournal Entry Account` other_line
                ON je.name = other_line.parent
                AND other_line.name != cash_line.name

            WHERE
                cash_line.account = %(account)s
                AND je.posting_date >= %(from_date)s

            GROUP BY
                je.name,
                cash_line.name

            ORDER BY je.creation DESC
            LIMIT 10
        )

        SELECT
            'Opening' AS voucher,
            %(from_date)s AS posting_date,
            '' AS creation,
            %(account)s AS cash_account,
            '' AS counter_details,
            'Opening Balance' AS remark,
            '' AS status,
            '' AS custom_payment_type,
            0 AS debit,
            0 AS credit,
            opening_balance AS running_balance
        FROM opening

        UNION ALL

        SELECT * FROM main

        ORDER BY posting_date, voucher

    """, {
        "account": account,
        "from_date": from_date
    }, as_dict=True)

    # ⭐ TOTAL CALCULATION
    total_debit = sum(d.debit for d in data)
    total_credit = sum(d.credit for d in data)

    return {
        "rows": data,
        "total_debit": total_debit,
        "total_credit": total_credit
    }

File: bfl_custom/bfl_custom/page/latest_journal_entry/latest_journal_entry.js
frappe.pages['latest_journal_entry'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'None',
		single_column: true
	});
	
	let account = page.add_field({
        label: 'Account',
        fieldtype: 'Link',
        fieldname: 'account',
        options: 'Account',
        reqd: 1
    });

    let from_date = page.add_field({
        label: 'From Date',
        fieldtype: 'Date',
        fieldname: 'from_date',
        reqd: 1
    });

    page.add_inner_button('Load', () => load_data());

    let $table = $('<div class="mt-4"></div>').appendTo(page.body);

    function load_data() {

        frappe.call({
            method: 'your_app.your_app.page.latest_journal_entry.latest_journal_entry.get_latest_journal_entries',
            args: {
                account: account.get_value(),
                from_date: from_date.get_value()
            },
            callback: function(r) {

                if (!r.message) return;

                render_table(r.message);
            }
        });
    }

    function render_table(res) {

        let data = res.rows;

        let html = `
            <table class="table table-bordered">
                <thead>
                    <tr>
                        <th>Voucher</th>
                        <th>Posting Date</th>
                        <th>Account</th>
                        <th>Counter Details</th>
                        <th>Remark</th>
                        <th>Status</th>
                        <th>Payment Type</th>
                        <th>Debit</th>
                        <th>Credit</th>
                        <th>Running Balance</th>
                    </tr>
                </thead>
                <tbody>
        `;

        data.forEach(row => {

            html += `
                <tr>
                    <td>${row.voucher}</td>
                    <td>${row.posting_date || ''}</td>
                    <td>${row.cash_account || ''}</td>
                    <td>${row.counter_details || ''}</td>
                    <td>${row.remark || ''}</td>
                    <td>${row.status || ''}</td>
                    <td>${row.custom_payment_type || ''}</td>
                    <td>${row.debit || 0}</td>
                    <td>${row.credit || 0}</td>
                    <td>${row.running_balance || 0}</td>
                </tr>
            `;
        });

        // ⭐ TOTAL ROW
        html += `
            <tr style="font-weight:bold; background:#f5f5f5;">
                <td colspan="7" style="text-align:right;">TOTAL</td>
                <td>${res.total_debit}</td>
                <td>${res.total_credit}</td>
                <td></td>
            </tr>
        `;

        html += `</tbody></table>`;

        $table.html(html);
    }
};